Add Azure Service Bus Queue and Subscription triggers for async message processing

prdai opened the pull request:
Add Azure Service Bus Queue and Subscription triggers

This PR adds two new triggers for monitoring Azure Service Bus:
- `AzureServiceBusQueueTrigger` - monitors Service Bus queues for messages
- `AzureServiceBusSubscriptionTrigger` - monitors topic subscriptions for messages

These triggers enable event-driven DAG execution when messages arrive in Azure Service Bus, supporting async message processing workflows.

## Changes
- New trigger classes with async message processing
- Base trigger class for shared functionality  
- Documentation with usage examples
- Comprehensive test coverage

## Testing
- Unit tests for both trigger classes
- Integration scenarios and edge cases
- Async test patterns with proper mocking

related: AIP-82 (#52712)

prdai committed:
Add Azure Service Bus integration with triggers and documentation

prdai committed:
Implement Azure Service Bus Queue and Subscription triggers with async message processing and unit tests

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

jason810496 reviewed (commented):
Nice! Thanks for the PR.

Would it be simpler if the `BaseAzureServiceBusTrigger` just raised the `TriggerEvent` with the `service_bus_client` directly?
I understand the idea behind using `asyncio.Queue` here, but IMO, we don't need to be constrained by the current methods available in `MessageHook` or use an `asyncio.Queue` just to accommodate the `MessageHook` interface.

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
feat(Azure): Add methods to read messages from Service Bus queue and subscription

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

jason810496 commented on providers/microsoft/azure/docs/connections/message_bus.rst:
```diff
@@ -0,0 +1,136 @@
+The ``queues`` parameter is used to configure the underlying
+:class:`~airflow.providers.microsoft.azure.triggers.message_bus.AzureServiceBusQueueTrigger` class.
+
+Queues can also be specified via the Queue URI instead of the ``queues`` kwarg. The provider will extract queues from the URI.
```
It seems this documentation should not be included in this PR.

jason810496 commented on providers/microsoft/azure/docs/connections/message_bus.rst:
```diff
@@ -0,0 +1,136 @@
+:class:`~airflow.providers.microsoft.azure.triggers.message_bus.AzureServiceBusSubscriptionTrigger` class.
+
+Topics can also be specified via the Queue URI instead of the ``topics`` kwarg. The provider will extract topics from the URI.
+``subscription_name`` must be specified in the URI.
```
I think this part should be excluded in this one as well.

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,186 @@
+
+import asyncio
+from collections.abc import AsyncGenerator
+
```
```suggestion
from typing import Any
```

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,186 @@
+        super().__init__(poll_interval, azure_service_bus_conn_id, max_wait_time)
+        self.queues = queues
+
+    def serialize(self) -> tuple[str, dict[str, any]]:
```
```suggestion
    def serialize(self) -> tuple[str, dict[str, Any]]:
```

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,186 @@
+        self.topics = topics
+        self.subscription_name = subscription_name
+
+    def serialize(self) -> tuple[str, dict[str, any]]:
```
```suggestion
    def serialize(self) -> tuple[str, dict[str, Any]]:
```

jason810496 reviewed (commented):
Nice! LGTM overall.
It seems some part of the documentation should be covered in further Queue Provider PR.

Just FYI, the "Queue URI" should be replace by "scheme" in further Queue Provider PR.
You may refer to https://github.com/apache/airflow/pull/54651/commits/5c03795b424d40d4d405c0c43aa93f53ab746236 as an example of how to adapt to “scheme”.

prdai committed:
Update providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

prdai committed:
Update providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

prdai committed:
Update providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
Remove Queue URI format documentation for Azure Service Bus Queue and Subscription Providers

jason810496 reviewed (commented):
The CI failure still need to be resolved, thanks!

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
Enhance Azure Service Bus integration by adding new message types and updating trigger initialization tests

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

jason810496 reviewed (commented):
I think we need to fix unit test instead of merging main to resolve the CI failure, thanks!

https://github.com/apache/airflow/actions/runs/17449765741/job/49552404061?pr=53356

prdai committed:
Fix Azure Service Bus message trigger test failures

- Simplified mocking to avoid connection and import issues
- Fixed mock assertions for MessageHook initialization
- Addressed compatibility issues with older Airflow versions
- Used context managers for proper mock lifecycle management

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
Fix Azure Service Bus trigger test mocking issues

  - Replace Mock object string representations with proper Mock(body=value) pattern
  - Fix str(message.body) calls returning mock object strings instead of expected message content
  - Apply standard Azure provider mocking conventions consistent with other tests
  - Remove unnecessary comments and custom MockMessage classes

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/hooks/asb.py:
```diff
@@ -606,6 +606,74 @@ def receive_subscription_message(
+            if received_msgs:
+                msg = received_msgs[0]
+                receiver.complete_message(msg)
+                return msg
```
Not sure do I understand correctly. Does this means that even we set `max_message_count` more than `1`, we will just complete `1` message intentionally?

prdai replied:
i think we do not need to actually have a parameter for `max_message_count` since we want to just get a singular message

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/hooks/asb.py:
```diff
@@ -606,6 +606,74 @@ def receive_subscription_message(
+            received_msgs = subscription_receiver.receive_messages(
+                max_message_count=max_message_count or 1, max_wait_time=max_wait_time
+            )
+            if received_msgs:
+                msg = received_msgs[0]
+                subscription_receiver.complete_message(msg)
+                return msg
+            return None
```
Same question here.

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,196 @@
+                    queue_name=queue_name, max_wait_time=self.max_wait_time
+                )
+                if message:
+                    yield TriggerEvent({"message": str(message.body), "queue": queue_name})
```
Are we able to deserialize back the `str(message.body)` ?

jason810496 reviewed (commented):
Nice! Thanks for the update! LGTM overall.

prdai committed:
Update the Trigger Event Message to Include the Actual Content

prdai committed:
Merge remote-tracking branch 'upstream/main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
resolve review comments

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

vincbeck commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,214 @@
+    from azure.servicebus import ServiceBusReceivedMessage
+
+
+class BaseAzureServiceBusTrigger(BaseTrigger):
```
All triggers that are compatible with event-driven scheduling should inherit `BaseEventTrigger `. Please look at this PR: https://github.com/apache/airflow/pull/56741. I am doing the same for Kafka

vincbeck reviewed (commented):
It could be done as a follow-up PR, but it would be also nice to add these 2 services as part of the unified `MessageQueueTrigger `. Please read [documentation](https://airflow.apache.org/docs/apache-airflow-providers-common-messaging/stable/providers.html) on how to do it.

The goal is to have one `MessageQueueTrigger` that is compatible with different queue providers, today we have SQS and Kafka. Would be awesome to add Azure as well

prdai committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

prdai committed:
chore: change BaseTrigger to BaseEventTrigger

jason810496 commented on providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py:
```diff
@@ -0,0 +1,214 @@
+from asgiref.sync import sync_to_async
+
+from airflow.providers.microsoft.azure.hooks.asb import MessageHook
+from airflow.triggers.base import BaseEventTrigger, TriggerEvent
```
Adding compact import _should_ fix the CI.

```suggestion
if AIRFLOW_V_3_0_PLUS:
    from airflow.triggers.base import BaseEventTrigger, TriggerEvent
else:
    from airflow.triggers.base import (  # type: ignore
        BaseTrigger as BaseEventTrigger,
        TriggerEvent,
    )
```

prdai replied:
thanks man! :)

jason810496 replied:
We still need to add `from tests_common.test_utils.version_compat import AIRFLOW_V_3_0_PLUS` to resolve the current error. Thanks!

prdai replied:
done :)

jason810496 reviewed (commented):
Thanks for the update!
LGTM if the CI pass.

> hi, can we take it up as a follow up PR?

Sure! We definitely can take this one just for adding Trigger of Azure Service Bus Queue, then the `MessageQueueTrigger` can be done in follow-up one.

prdai committed:
Update providers/microsoft/azure/src/airflow/providers/microsoft/azure/triggers/message_bus.py

Co-authored-by: LIU ZHE YOU <68415893+jason810496@users.noreply.github.com>

prdai committed:
chore: fix ci/cd

jason810496 committed:
Merge branch 'main' into ffr/aip-82-add-azure-message-bus-integration

potiuk reviewed (commented):
LGTM. I have no big experience with Azure Service Bus but it looks **sound** and there are no red flags.
